crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -306,11 +322,133 @@ impl SnapshotRepository {
+                info!("Latest snapshot points to a snapshot that was not found in the repository!");
+                return Ok(None); // arguably this could also be an error
+            }
+            Err(e) => return Err(e.into()),
+        };
+
+        let mut snapshot_metadata: PartitionSnapshotMetadata =
+            serde_json::from_slice(snapshot_metadata.bytes().await?.iter().as_slice())?;
+        if snapshot_metadata.version != SnapshotFormatVersion::V1 {
+            return Err(anyhow!(
+                "Unsupported snapshot format version: {:?}",
+                snapshot_metadata.version
+            ));
+        }
+
+        // The snapshot ingest directory should be on the same filesystem as the partition store
+        // to minimize IO and disk space usage during import.
+        let snapshot_dir = TempDir::with_prefix_in(
+            format!("{}-", snapshot_metadata.snapshot_id),
+            &self.base_dir,
+        )?;
+        debug!(
+            snapshot_id = %snapshot_metadata.snapshot_id,
+            path = ?snapshot_dir.path(),
+            "Getting snapshot data",
+        );
+
+        let directory = snapshot_dir.path().to_string_lossy().to_string();
+        let concurrency_limiter = Arc::new(Semaphore::new(DOWNLOAD_CONCURRENCY_LIMIT));
+        let mut downloads = JoinSet::new();
+        for file in &mut snapshot_metadata.files {
+            let filename = file.name.trim_start_matches("/");
+            let key = object_store::path::Path::from(format!(
+                "{prefix}{partition_id}/{path}/{filename}",
+                prefix = self.prefix,
+                partition_id = partition_id,
+                path = latest.path,
+                filename = filename,
+            ));
+            let file_path = snapshot_dir.path().join(filename);
+            let concurrency_limiter = Arc::clone(&concurrency_limiter);
+            let object_store = Arc::clone(&self.object_store);
+
+            downloads.spawn(async move {
+                let _permit = concurrency_limiter.acquire().await?;
+                let mut file_data = StreamReader::new(object_store.get(&key).await?.into_stream());
+                let mut snapshot_file = tokio::fs::File::create_new(&file_path).await?;
+                let size = io::copy(&mut file_data, &mut snapshot_file).await?;
+                trace!(%key, ?size, "Downloaded snapshot data file to {:?}", file_path);
+                anyhow::Ok(())
+            });
+            // patch the directory path to reflect the actual location on the restoring node
+            file.directory = directory.clone();
+        }
+
+        loop {
+            match downloads.join_next().await {
+                None => {
+                    break;
+                }
+                Some(Err(e)) => {
+                    downloads.abort_all();
+                    return Err(e.into());
+                }
+                Some(Ok(_)) => {}
```

Comment:
Is it intentional to not handle errors returned by the download routine ? the `_` here is actually an `anyhow::Result` which itself can be an error.

Check suggestion below

Reply:
This is a pretty big miss, thanks for catching it!

Reply:
Updated, let me know how you like it! I might have overdone the error handling a bit 😅